File: DataSets/DataSet-Old/LeetCodeTasks - Develop/LeetCodeTasks/63_unique-paths-ii/solutions/cpp/solution.cpp
#include <vector>

class Solution {
public:
    int uniquePathsWithObstacles(std::vector<std::vector<int>>& obstacleGrid) {
        if (obstacleGrid.empty() || obstacleGrid[0][0] == 1) return 0;

        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        std::vector<std::vector<long>> dp(m, std::vector<long>(n, 0));

        dp[0][0] = obstacleGrid[0][0] == 0 ? 1 : 0;

        for (int i = 1; i < m; ++i) {
            if (obstacleGrid[i][0] == 0 && dp[i-1][0] == 1)
                dp[i][0] = 1;
        }
        for (int j = 1; j < n; ++j) {
            if (obstacleGrid[0][j] == 0 && dp[0][j-1] == 1)
                dp[0][j] = 1;
        }

        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                if (obstacleGrid[i][j] == 0) {
                    dp[i][j] = dp[i-1][j] + dp[i][j-1];
                }
            }
        }

        return dp[m-1][n-1];
    }
};

File: DataSets/DataSet-Old/LeetCodeTasks - Develop/LeetCodeTasks/63_unique-paths-ii/solutions/python3/solution.py
from typing import List

class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        # If the starting or ending point is an obstacle, return 0 paths.
        if obstacleGrid[0][0] == 1 or obstacleGrid[m-1][n-1] == 1:
            return 0
        
        # Initialize the dynamic programming grid with zeros.
        dp = [[0] * n for _ in range(m)]
        # Starting point has 1 way if not an obstacle.
        dp[0][0] = 1
        
        # Fill the dp grid
        for i in range(m):
            for j in range(n):
                if obstacleGrid[i][j] == 1:
                    dp[i][j] = 0
                else:
                    if i > 0:
                        dp[i][j] += dp[i-1][j]
                    if j > 0:
                        dp[i][j] += dp[i][j-1]
                        
        return dp[m-1][n-1]
